### DISCUSSION TITLE
Cleaning up handling of units for alignment

### DISCUSSION THREAD

**Discussion by RickBrice** [IfcOpenShell/IfcOpenShell#5670] [General]
@aothms @civilx64 there has been a few issues around units of length measure and the new alignment geometry code. I wanted to have a quick discussion to make sure we are on the same page and to get the coding cleaned up.

All of the alignment geometry processing boils down to calling `Eigen::Matrix4d piecewise_function_evaluator.evaluate(double u)`.

1. Should `u` be in the length unit of the IFC model or meter? Currently expects length to be in meter.
2. Should the return value be in the length unit of the IFC model or meter? Current returns meter. To get (X,Y,Z) in model units, the caller must divided by the return values by `ifcopenshell::geometry::abstract_mapping::get_length_unit()`
3. Should the `PiecewiseStepParam` be in the length unit of the IFC model or meter? Current defaults to 0.5 and is taken to be in meter.

In the current implementation of `IfcCurveSegment`, SegmentStart, SegmentLength, and the parent curve parameters are converted to meter. The unit conversion was inherited from the early prototype code. The geometry calculations would work equally well if all they were done in the units of the IFC model, and unit conversion could be completely avoided.

I think these 3 questions are easily answered by understanding the broader approach to handing units in IFCOS. The alignment geometry should be consistent with current implementation for other geometries. If a cube is defined with FOOT units, when that geometry gets mapped to the general taxonomy are the resulting corners of the cube in FOOT or METER?

**Comment by civilx64** [comment]
> If a cube is defined with FOOT units, when that geometry gets mapped to the general taxonomy are the resulting corners of the cube in FOOT or METER?

The vertices come out in METER - which makes sense given that other output formats such as glTF require the geometry to be defined in meters.  So for consistency I think we should keep using meter for everything in the core C++ layer.  Then we can incorporate the unit conversion to and from IFC model units in the python layer.

**Comment by RickBrice** [comment]
Thank makes sense, and it is the current implementation so no changes. I like that!
Basically, it doesn't matter what units are defined in the model - all length inputs and outputs with IfcOpenShell are in METER.

With that in mind, f3ce57480a04246560ed268a5017e1e82c933fee fixes the unit conversion problems. I have the Rail Room models converted to FOOT units. Not sure where to dump them. Evandro is working on something for that. In the meantime, I can share by notebook scripts if you want to generate the files yourself for VS regression tests.
